Deploy on Heroku using Docker

silviot opened the pull request:
## Description
Provide ability to deploy to Heroku using docker containers. See #2556

Much of the diff is code moved to `utils.py` to prevent circular import issues.

## How Has This Been Tested?
Using `bartelet1932` the following steps work:

* `dallinger docker sandbox --verbose`
* `dallinger export`
* `dallinger destroy`

Some features are still missing, among which:
* properly scaling dynos
* support for experiment zip archive

pmcharrison commented:
`dallinger docker debug` called when the Docker Desktop service is not running produces this error message:

```
During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/peter.harrison/.virtualenvs/psynet/bin/dallinger", line 11, in <module>
    load_entry_point('dallinger', 'console_scripts', 'dallinger')()
  File "/Users/peter.harrison/.virtualenvs/psynet/lib/python3.9/site-packages/click/core.py", line 829, in __call__
    return self.main(*args, **kwargs)
  File "/Users/peter.harrison/.virtualenvs/psynet/lib/python3.9/site-packages/click/core.py", line 782, in main
    rv = self.invoke(ctx)
  File "/Users/peter.harrison/.virtualenvs/psynet/lib/python3.9/site-packages/click/core.py", line 1259, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/Users/peter.harrison/.virtualenvs/psynet/lib/python3.9/site-packages/click/core.py", line 1259, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/Users/peter.harrison/.virtualenvs/psynet/lib/python3.9/site-packages/click/core.py", line 1066, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/Users/peter.harrison/.virtualenvs/psynet/lib/python3.9/site-packages/click/core.py", line 610, in invoke
    return callback(*args, **kwargs)
  File "/Users/peter.harrison/gdrive/academic/projects/harrison-peter/Dallinger/dallinger/command_line/utils.py", line 72, in wrapper
    return f(**kwargs)
  File "/Users/peter.harrison/gdrive/academic/projects/harrison-peter/Dallinger/dallinger/command_line/docker.py", line 44, in debug
    debugger.run()
  File "/Users/peter.harrison/gdrive/academic/projects/harrison-peter/Dallinger/dallinger/deployment.py", line 703, in run
    with self.WRAPPER_CLASS(
  File "/Users/peter.harrison/gdrive/academic/projects/harrison-peter/Dallinger/dallinger/docker/tools.py", line 56, in __init__
    self.client = docker.APIClient(base_url="unix://var/run/docker.sock")
  File "/Users/peter.harrison/.virtualenvs/psynet/lib/python3.9/site-packages/docker/api/client.py", line 197, in __init__
    self._version = self._retrieve_server_version()
  File "/Users/peter.harrison/.virtualenvs/psynet/lib/python3.9/site-packages/docker/api/client.py", line 221, in _retrieve_server_version
    raise DockerException(
docker.errors.DockerException: Error while fetching server API version: ('Connection aborted.', FileNotFoundError(2, 'No such file or directory'))
```

I wonder if there is an easy way to make a more informative error message in this case?

pmcharrison commented:
`dallinger docker debug` works well for me when Dallinger is installed as editable. When installed without the editable flag, I get the following error on various demos:

```

#7 [4/4] RUN bash -c "python3 -m pip install -r <(grep -v ^dallinger requir...
#7 CACHED

#9 exporting to image
#9 exporting layers done
#9 writing image sha256:6cc6ccf744a9e0985a6406e623341f32519a5099e06d2d1a723cc98de3b23cbc done
#9 naming to docker.io/library/mcmcp-web done
#9 DONE 0.0s
Starting mcmcp_redis_1      ... done
Recreating mcmcp_web_1      ... done
Starting mcmcp_postgresql_1 ... done
Recreating mcmcp_worker_1   ... done
Postgresql ready
Redis ready

❯❯ Some services did not start properly:

❯❯ mcmcp_web_1: exited - healthy
Traceback (most recent call last):
  File "/Users/peter.harrison/.virtualenvs/psynet/lib/python3.9/site-packages/docker/api/client.py", line 268, in _raise_for_status
    response.raise_for_status()
  File "/Users/peter.harrison/.virtualenvs/psynet/lib/python3.9/site-packages/requests/models.py", line 943, in raise_for_status
    raise HTTPError(http_error_msg, response=self)
requests.exceptions.HTTPError: 404 Client Error: Not Found for url: http+docker://localhost/v1.40/containers/mcmcp_web_1:/json

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/peter.harrison/.virtualenvs/psynet/bin/dallinger", line 8, in <module>
    sys.exit(dallinger())
  File "/Users/peter.harrison/.virtualenvs/psynet/lib/python3.9/site-packages/click/core.py", line 829, in __call__
    return self.main(*args, **kwargs)
  File "/Users/peter.harrison/.virtualenvs/psynet/lib/python3.9/site-packages/click/core.py", line 782, in main
    rv = self.invoke(ctx)
  File "/Users/peter.harrison/.virtualenvs/psynet/lib/python3.9/site-packages/click/core.py", line 1259, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/Users/peter.harrison/.virtualenvs/psynet/lib/python3.9/site-packages/click/core.py", line 1259, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/Users/peter.harrison/.virtualenvs/psynet/lib/python3.9/site-packages/click/core.py", line 1066, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/Users/peter.harrison/.virtualenvs/psynet/lib/python3.9/site-packages/click/core.py", line 610, in invoke
    return callback(*args, **kwargs)
  File "/Users/peter.harrison/.virtualenvs/psynet/lib/python3.9/site-packages/dallinger/command_line/utils.py", line 72, in wrapper
    return f(**kwargs)
  File "/Users/peter.harrison/.virtualenvs/psynet/lib/python3.9/site-packages/dallinger/command_line/docker.py", line 44, in debug
    debugger.run()
  File "/Users/peter.harrison/.virtualenvs/psynet/lib/python3.9/site-packages/dallinger/deployment.py", line 703, in run
    with self.WRAPPER_CLASS(
  File "/Users/peter.harrison/.virtualenvs/psynet/lib/python3.9/site-packages/dallinger/docker/tools.py", line 81, in __enter__
    return self.start()
  File "/Users/peter.harrison/.virtualenvs/psynet/lib/python3.9/site-packages/dallinger/docker/tools.py", line 145, in start
    self.client.attach(error.split(" ")[0], logs=True).decode("utf-8")
  File "/Users/peter.harrison/.virtualenvs/psynet/lib/python3.9/site-packages/docker/utils/decorators.py", line 19, in wrapped
    return f(self, resource_id, *args, **kwargs)
  File "/Users/peter.harrison/.virtualenvs/psynet/lib/python3.9/site-packages/docker/api/container.py", line 63, in attach
    response, stream, self._check_is_tty(container), demux=demux)
  File "/Users/peter.harrison/.virtualenvs/psynet/lib/python3.9/site-packages/docker/utils/decorators.py", line 19, in wrapped
    return f(self, resource_id, *args, **kwargs)
  File "/Users/peter.harrison/.virtualenvs/psynet/lib/python3.9/site-packages/docker/api/client.py", line 454, in _check_is_tty
    cont = self.inspect_container(container)
  File "/Users/peter.harrison/.virtualenvs/psynet/lib/python3.9/site-packages/docker/utils/decorators.py", line 19, in wrapped
    return f(self, resource_id, *args, **kwargs)
  File "/Users/peter.harrison/.virtualenvs/psynet/lib/python3.9/site-packages/docker/api/container.py", line 773, in inspect_container
    return self._result(
  File "/Users/peter.harrison/.virtualenvs/psynet/lib/python3.9/site-packages/docker/api/client.py", line 274, in _result
    self._raise_for_status(response)
  File "/Users/peter.harrison/.virtualenvs/psynet/lib/python3.9/site-packages/docker/api/client.py", line 270, in _raise_for_status
    raise create_api_error_from_http_exception(e)
  File "/Users/peter.harrison/.virtualenvs/psynet/lib/python3.9/site-packages/docker/errors.py", line 31, in create_api_error_from_http_exception
    raise cls(e, response=response, explanation=explanation)
docker.errors.NotFound: 404 Client Error for http+docker://localhost/v1.40/containers/mcmcp_web_1:/json: Not Found ("No such container: mcmcp_web_1:")
```

Do you see this too, `silviot`?

pmcharrison commented:
It seems that whenever I run `dallinger docker debug`, it builds a new image, is that right? I thought maybe this would be cached?

pmcharrison commented:
Using git in requirements.txt causes an error:

```
#8 [4/4] RUN bash -c "python3 -m pip install -r <(grep -v ^dallinger requir...
#8 1.109 Obtaining psynet from git+https://psynet:****@gitlab.com/computational-audition-lab/psynet#egg=psynet (from -r /dev/fd/63 (line 9))
#8 1.110   Cloning https://psynet:****@gitlab.com/computational-audition-lab/psynet to ./src/psynet
#8 1.110   Running command git clone -q 'https://psynet:****@gitlab.com/computational-audition-lab/psynet' /experiment/src/psynet
#8 1.112   ERROR: Error [Errno 2] No such file or directory: 'git' while executing command git clone -q 'https://psynet:****@gitlab.com/computational-audition-lab/psynet' /experiment/src/psynet
#8 1.113 ERROR: Cannot find command 'git' - do you have 'git' installed and in your PATH?
#8 ERROR: executor failed running [/bin/sh -c bash -c "python3 -m pip install -r <(grep -v ^dallinger requirements.txt)"]: runc did not terminate sucessfully
```

Maybe we should include git in the base image?

silviot commented:
@pmcharrison The base image does not include git, since it would then have to be always present and would increase upload times of experiments for everyone.
You can place a file `prepare_docker_image.sh` in your experiment directory with this content:

```
#!/bin/sh

apt-get update
DEBIAN_FRONTEND=noninteractive apt-get install -y git
# Save space taken by the final image by removing unnecessary files
rm -rf /var/lib/apt/lists/*
```
and it will install git before installing dependencies.
Also make sure you run `dallinger generate-constraints` to make sure you're using a current version of the base docker image.

`dallinger docker debug` should indeed generate a new image only if  `prepare_docker_image.sh` or `constraints.txt` changed.

I added a more graceful message when the docker daemon is not running. 

I rebased this branch against `master` so you'll need

```
git fetch origin issue/2556-heroku-docker-deploymemnt
git reset --hard origin/issue/2556-heroku-docker-deploymemnt
```
to test it again.

I was not able to reproduce the issue you mention when installing in non editable mode, but I believe it's related to the test failures, and that both are related to the new `disable_when_duration_exceeded` parameter. I'm still investigating.

pmcharrison commented:
Hi @silviot, thanks for your help. I'm trying the new branch now with Dallinger installed in editable mode. I get the following error:

```
Traceback (most recent call last):
  File "/Users/peter.harrison/.virtualenvs/psynet/bin/dallinger", line 11, in <module>
    load_entry_point('dallinger', 'console_scripts', 'dallinger')()
  File "/Users/peter.harrison/.virtualenvs/psynet/lib/python3.9/site-packages/click/core.py", line 829, in __call__
    return self.main(*args, **kwargs)
  File "/Users/peter.harrison/.virtualenvs/psynet/lib/python3.9/site-packages/click/core.py", line 782, in main
    rv = self.invoke(ctx)
  File "/Users/peter.harrison/.virtualenvs/psynet/lib/python3.9/site-packages/click/core.py", line 1259, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/Users/peter.harrison/.virtualenvs/psynet/lib/python3.9/site-packages/click/core.py", line 1259, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/Users/peter.harrison/.virtualenvs/psynet/lib/python3.9/site-packages/click/core.py", line 1066, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/Users/peter.harrison/.virtualenvs/psynet/lib/python3.9/site-packages/click/core.py", line 610, in invoke
    return callback(*args, **kwargs)
  File "/Users/peter.harrison/gdrive/academic/projects/harrison-peter/Dallinger/dallinger/command_line/utils.py", line 72, in wrapper
    return f(**kwargs)
  File "/Users/peter.harrison/gdrive/academic/projects/harrison-peter/Dallinger/dallinger/command_line/docker.py", line 65, in debug
    debugger.run()
  File "/Users/peter.harrison/gdrive/academic/projects/harrison-peter/Dallinger/dallinger/deployment.py", line 284, in run
    with self.WRAPPER_CLASS(
  File "/Users/peter.harrison/gdrive/academic/projects/harrison-peter/Dallinger/dallinger/docker/tools.py", line 89, in __enter__
    return self.start()
  File "/Users/peter.harrison/gdrive/academic/projects/harrison-peter/Dallinger/dallinger/docker/tools.py", line 115, in start
    build_image(self.tmp_dir, self.experiment_name, self.out)
  File "/Users/peter.harrison/gdrive/academic/projects/harrison-peter/Dallinger/dallinger/docker/tools.py", line 250, in build_image
    base_image_name = get_base_image(tmp_dir)
  File "/Users/peter.harrison/gdrive/academic/projects/harrison-peter/Dallinger/dallinger/docker/tools.py", line 205, in get_base_image
    dallinger_version = get_required_dallinger_version(experiment_tmp_path)
  File "/Users/peter.harrison/gdrive/academic/projects/harrison-peter/Dallinger/dallinger/docker/tools.py", line 224, in get_required_dallinger_version
    return dallinger_requirements[0].split("==")[1]
IndexError: list index out of range
```

If I look in the temporary directory I see the following entry for Dallinger:

```
file:dallinger-7.2.0-py3-none-any.whl
    # via -r requirements.txt
```

It seems like `get_required_dallinger_version` fails with this kind of format. Any ideas about how to move forward here?

In my experiment directory requirements.txt I had this:

```
dallinger==7.2.0
```

pmcharrison commented:
Also seeing a lot of this deprecation warning:

```
DeprecationWarning: Creating a LegacyVersion has been deprecated and will be removed in the next major release
```

pmcharrison commented:
The 'docker is not running' error message works great, thanks!

pmcharrison commented:
The caching works great, I see it now, thanks!

I almost got there with `dallinger docker debug`. At the last point though I got this:

```
❯❯ Error accessing http://localhost:5000/launch (500):
{"status": "error", "message": "Failed to load experiment in /launch: No experiment classes found"}

❯❯ Experiment launch failed, check web dyno logs for details.

❯❯ Failed to load experiment in /launch: No experiment classes found
```

When I entered the container, `experiment.py` looked normal, though.

I wonder, @silviot, would you be willing to try yourself to get one of our demos working? I will add you to the PsyNet GitLab repository if you give me your email address. Maybe you'd be able to make the final steps, instead of painful remote debugging.

silviot commented:
@pmcharrison Sure! My gitlab account has email silviot@gmail.com.

pmcharrison commented:
Thanks so much! Have added you, the merge request is here: https://gitlab.com/computational-audition-lab/psynet/-/merge_requests/145

silviot commented:
@pmcharrison With the latest changes I was able to deploy the psynet demo to Heroku using docker.
Both the `dallinger docker sandbox` and `dallinger docker deploy` commands should work now. 
Dynos should be scaled according to `config.txt`.
Can you give it a try?

pmcharrison commented:
Hi @silviot, sorry, a little confused - the current requirements.txt in the mcmcp demo has `dallinger==7.2.0`. But then how does it access your latest Dallinger changes? How are you running it? 

I tried putting the following in requirements.txt:

```
-e git+https://github.com/dallinger/dallinger@issue/2556-heroku-docker-deploymemnt#egg=dallinger
file:psynet-1.12.0-py3-none-any.whl
```

and ran `dallinger generate-constraints`, then `dallinger docker debug`, but then got this error:

```
Could not determine Dallinger version. Using latest
Image mcmcp:a4e45d2a not found - building
Step 1/5 : FROM ghcr.io/dallinger/dallinger:
Traceback (most recent call last):
  File "/Users/peter.harrison/.virtualenvs/psynet/bin/dallinger", line 8, in <module>
    sys.exit(dallinger())
  File "/Users/peter.harrison/.virtualenvs/psynet/lib/python3.9/site-packages/click/core.py", line 829, in __call__
    return self.main(*args, **kwargs)
  File "/Users/peter.harrison/.virtualenvs/psynet/lib/python3.9/site-packages/click/core.py", line 782, in main
    rv = self.invoke(ctx)
  File "/Users/peter.harrison/.virtualenvs/psynet/lib/python3.9/site-packages/click/core.py", line 1259, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/Users/peter.harrison/.virtualenvs/psynet/lib/python3.9/site-packages/click/core.py", line 1259, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/Users/peter.harrison/.virtualenvs/psynet/lib/python3.9/site-packages/click/core.py", line 1066, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/Users/peter.harrison/.virtualenvs/psynet/lib/python3.9/site-packages/click/core.py", line 610, in invoke
    return callback(*args, **kwargs)
  File "/Users/peter.harrison/.virtualenvs/psynet/lib/python3.9/site-packages/dallinger/command_line/utils.py", line 72, in wrapper
    return f(**kwargs)
  File "/Users/peter.harrison/.virtualenvs/psynet/lib/python3.9/site-packages/dallinger/command_line/docker.py", line 65, in debug
    debugger.run()
  File "/Users/peter.harrison/.virtualenvs/psynet/lib/python3.9/site-packages/dallinger/deployment.py", line 284, in run
    with self.WRAPPER_CLASS(
  File "/Users/peter.harrison/.virtualenvs/psynet/lib/python3.9/site-packages/dallinger/docker/tools.py", line 90, in __enter__
    return self.start()
  File "/Users/peter.harrison/.virtualenvs/psynet/lib/python3.9/site-packages/dallinger/docker/tools.py", line 116, in start
    build_image(self.tmp_dir, self.experiment_name, self.out)
  File "/Users/peter.harrison/.virtualenvs/psynet/lib/python3.9/site-packages/dallinger/docker/tools.py", line 314, in build_image
    raise BuildError(line_decoded["error"])
dallinger.docker.tools.BuildError: invalid reference format
```

Would you mind hazarding a guess about what I'm doing wrong?

silviot commented:
@pmcharrison I believe there was a bug when unable to determine the dallinger version from the `requirements.txt`. Now I pushed a fix so that the latest version is used in that case.

silviot commented:
> Hi @silviot, sorry, a little confused - the current requirements.txt in the mcmcp demo has `dallinger==7.2.0`. But then how does it access your latest Dallinger changes? How are you running it?

You'll need to install dallinger in editable mode from this branch to make it work. This is meant to be the preferred way of testing dallinger code changes when using it with docker.

If you still have problems can you paste here the output of `grep dallinger== demos/mcmcp/constraints.txt` and  'pip --version '?
Also, maybe installing the same pip version I have might help: `pip install pip==20.2.3`.

pmcharrison commented:
Thanks @silviot, makes sense! `dallinger docker debug` worked great now, thank you!

On `dallinger docker deploy` I got this:

```
❯❯ Making sure that you are logged in to Heroku.
 ›   Warning: token will expire 08/11/2021
 ›   Use heroku authorizations:create to generate a long-term token

Creating dlgr-phtest... done
Setting CREATOR, DALLINGER_UID, HOST and restarting dlgr-phtest... done, v3
Creating heroku-postgresql:standard-2 on dlgr-phtest... $200/month
Creating heroku-redis:hobby-dev on dlgr-phtest... free
Creating papertrail on dlgr-phtest... free
Setting FLASK_SECRET_KEY, auto_recruit, aws_access_key_id, aws_region, aws_secret_access_key, dashboard_password, smtp_password, smtp_username, whimsical and restarting dlgr-phtest... done, v6
no basic auth credentials
 ▸    Error: docker push exited with Error: 1
```

Is this perhaps referring to docker registry credentials?

silviot commented:
@pmcharrison I should add a check that suggests to run `heroku container:login` in case you haven't already, just as the one for Heroku login.

pmcharrison commented:
Good idea to add a check for `heroku container:login`, thanks! It got me past that error. I then saw this:

```
❯❯ Waiting for addons to be ready...
script: illegal option -- -
usage: script [-adkpqr] [-t time] [file [command ...]]
Traceback (most recent call last):
  File "/Users/peter.harrison/.virtualenvs/psynet/bin/dallinger", line 11, in <module>
    load_entry_point('dallinger', 'console_scripts', 'dallinger')()
  File "/Users/peter.harrison/.virtualenvs/psynet/lib/python3.9/site-packages/click/core.py", line 829, in __call__
    return self.main(*args, **kwargs)
  File "/Users/peter.harrison/.virtualenvs/psynet/lib/python3.9/site-packages/click/core.py", line 782, in main
    rv = self.invoke(ctx)
  File "/Users/peter.harrison/.virtualenvs/psynet/lib/python3.9/site-packages/click/core.py", line 1259, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/Users/peter.harrison/.virtualenvs/psynet/lib/python3.9/site-packages/click/core.py", line 1259, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/Users/peter.harrison/.virtualenvs/psynet/lib/python3.9/site-packages/click/core.py", line 1066, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/Users/peter.harrison/.virtualenvs/psynet/lib/python3.9/site-packages/click/core.py", line 610, in invoke
    return callback(*args, **kwargs)
  File "/Users/peter.harrison/gdrive/academic/projects/harrison-peter/Dallinger/dallinger/command_line/utils.py", line 72, in wrapper
    return f(**kwargs)
  File "/Users/peter.harrison/gdrive/academic/projects/harrison-peter/Dallinger/dallinger/command_line/docker.py", line 98, in sandbox
    _deploy_in_mode(mode="sandbox", verbose=verbose, log=log, app=app)
  File "/Users/peter.harrison/gdrive/academic/projects/harrison-peter/Dallinger/dallinger/command_line/docker.py", line 120, in _deploy_in_mode
    deploy_heroku_docker(log=log, verbose=verbose, app=app, prelaunch_actions=prelaunch)
  File "/Users/peter.harrison/gdrive/academic/projects/harrison-peter/Dallinger/dallinger/command_line/docker.py", line 217, in deploy_heroku_docker
    addons_text = heroku_app._result(heroku_addons_cmd(heroku_app.name)).strip()
  File "/Users/peter.harrison/gdrive/academic/projects/harrison-peter/Dallinger/dallinger/heroku/tools.py", line 80, in _result
    output = check_output(cmd)
  File "/Users/peter.harrison/gdrive/academic/projects/harrison-peter/Dallinger/dallinger/utils.py", line 240, in wrapper
    return func(*popenargs, **kwargs)
  File "/Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/subprocess.py", line 424, in check_output
    return run(*popenargs, stdout=PIPE, timeout=timeout, check=True,
  File "/Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/subprocess.py", line 528, in run
    raise CalledProcessError(retcode, process.args,
subprocess.CalledProcessError: Command '['script', '-q', '--command', 'heroku addons -a dlgr-phtest']' returned non-zero exit status 1.
```

This is with me running the command `dallinger docker sandbox --app phtest`, I also get a similar error if I simply run `dallinger docker sandbox`.

silviot commented:
@jcpeterson There was a bug in invoking `script` (I try to do it to keep the colors of the output of `heroku addons`).
I pushed a fix. Can you retry?

pmcharrison commented:
Works now - great! 

I did see an error in the logs but things seemed to proceed anyway:

```
Creating dlgr-phtest... done
Setting CREATOR, DALLINGER_UID, HOST and restarting dlgr-phtest... done, v3
Creating heroku-postgresql:standard-2 on dlgr-phtest... $200/month
Creating heroku-redis:hobby-dev on dlgr-phtest... free
Creating papertrail on dlgr-phtest... free
Setting FLASK_SECRET_KEY, auto_recruit, aws_access_key_id, aws_region, aws_secret_access_key, dashboard_password, smtp_password, smtp_username, whimsical and restarting dlgr-phtest... done, v6
Head https://heroku-docker-images-us.s3.amazonaws.com/docker/registry/v2/blobs/sha256/65/656ee4c3549197f387c75128611ceec97f9a7e38ed41c6d13ee8b2b6f74a9dfd/data?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=AKIAJRVCXVACJ2NKZF6A%2F20210415%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20210415T084212Z&X-Amz-Expires=1200&X-Amz-SignedHeaders=host&X-Amz-Signature=9c1ffc3739339d371b9b655a4bdfe4d664388c13dedcd3f820915f83921b6494: EOF
 ▸    Error: docker push exited with Error: 1
```

I also see the illegal option error on other dallinger commands, e.g.

```
peter.harrison@PCX0117 mcmcp % dallinger expire --sandbox --app phtest 
script: illegal option -- -
usage: script [-adkpqr] [-t time] [file [command ...]]
```

pmcharrison commented:
I tested some more for a key use case of ours: using Docker to overcome the 'slug-size' limit of Heroku Git deploy. The particular case study is adding a hefty Python package to requirements.txt; I added `tensorflow` to the MCMCP demo we've been trying previously. This more or less worked, which is great. However:

- The first time the docker image is pushed, it can take a long time, 11 minutes in our case (despite fast internet connection). There is no progress status in the console, and a user might therefore think that the app has crashed. I tried adding `--verbose` to the `heroku container:push` command but unfortunately this didn't achieve much. I'm not sure how best we could resolve this. I was impressed by the caching though, this worked very well.
- The call to `launch` seems to timeout:

```
❯❯ Launching the experiment on the remote server and starting recruitment...
Calling https://dlgr-phtest.herokuapp.com/launch
Traceback (most recent call last):
  File "/Users/peter.harrison/.virtualenvs/psynet/lib/python3.9/site-packages/urllib3/connectionpool.py", line 699, in urlopen
    httplib_response = self._make_request(
  File "/Users/peter.harrison/.virtualenvs/psynet/lib/python3.9/site-packages/urllib3/connectionpool.py", line 445, in _make_request
    six.raise_from(e, None)
  File "<string>", line 3, in raise_from
  File "/Users/peter.harrison/.virtualenvs/psynet/lib/python3.9/site-packages/urllib3/connectionpool.py", line 440, in _make_request
    httplib_response = conn.getresponse()
  File "/Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/http/client.py", line 1347, in getresponse
    response.begin()
  File "/Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/http/client.py", line 307, in begin
    version, status, reason = self._read_status()
  File "/Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/http/client.py", line 276, in _read_status
    raise RemoteDisconnected("Remote end closed connection without"
http.client.RemoteDisconnected: Remote end closed connection without response

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/peter.harrison/.virtualenvs/psynet/lib/python3.9/site-packages/requests/adapters.py", line 439, in send
    resp = conn.urlopen(
  File "/Users/peter.harrison/.virtualenvs/psynet/lib/python3.9/site-packages/urllib3/connectionpool.py", line 755, in urlopen
    retries = retries.increment(
  File "/Users/peter.harrison/.virtualenvs/psynet/lib/python3.9/site-packages/urllib3/util/retry.py", line 531, in increment
    raise six.reraise(type(error), error, _stacktrace)
  File "/Users/peter.harrison/.virtualenvs/psynet/lib/python3.9/site-packages/urllib3/packages/six.py", line 734, in reraise
    raise value.with_traceback(tb)
  File "/Users/peter.harrison/.virtualenvs/psynet/lib/python3.9/site-packages/urllib3/connectionpool.py", line 699, in urlopen
    httplib_response = self._make_request(
  File "/Users/peter.harrison/.virtualenvs/psynet/lib/python3.9/site-packages/urllib3/connectionpool.py", line 445, in _make_request
    six.raise_from(e, None)
  File "<string>", line 3, in raise_from
  File "/Users/peter.harrison/.virtualenvs/psynet/lib/python3.9/site-packages/urllib3/connectionpool.py", line 440, in _make_request
    httplib_response = conn.getresponse()
  File "/Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/http/client.py", line 1347, in getresponse
    response.begin()
  File "/Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/http/client.py", line 307, in begin
    version, status, reason = self._read_status()
  File "/Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/http/client.py", line 276, in _read_status
    raise RemoteDisconnected("Remote end closed connection without"
urllib3.exceptions.ProtocolError: ('Connection aborted.', RemoteDisconnected('Remote end closed connection without response'))

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/peter.harrison/.virtualenvs/psynet/bin/dallinger", line 11, in <module>
    load_entry_point('dallinger', 'console_scripts', 'dallinger')()
  File "/Users/peter.harrison/.virtualenvs/psynet/lib/python3.9/site-packages/click/core.py", line 829, in __call__
    return self.main(*args, **kwargs)
  File "/Users/peter.harrison/.virtualenvs/psynet/lib/python3.9/site-packages/click/core.py", line 782, in main
    rv = self.invoke(ctx)
  File "/Users/peter.harrison/.virtualenvs/psynet/lib/python3.9/site-packages/click/core.py", line 1259, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/Users/peter.harrison/.virtualenvs/psynet/lib/python3.9/site-packages/click/core.py", line 1259, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/Users/peter.harrison/.virtualenvs/psynet/lib/python3.9/site-packages/click/core.py", line 1066, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/Users/peter.harrison/.virtualenvs/psynet/lib/python3.9/site-packages/click/core.py", line 610, in invoke
    return callback(*args, **kwargs)
  File "/Users/peter.harrison/gdrive/academic/projects/harrison-peter/Dallinger/dallinger/command_line/utils.py", line 72, in wrapper
    return f(**kwargs)
  File "/Users/peter.harrison/gdrive/academic/projects/harrison-peter/Dallinger/dallinger/command_line/docker.py", line 98, in sandbox
    _deploy_in_mode(mode="sandbox", verbose=verbose, log=log, app=app)
  File "/Users/peter.harrison/gdrive/academic/projects/harrison-peter/Dallinger/dallinger/command_line/docker.py", line 120, in _deploy_in_mode
    deploy_heroku_docker(log=log, verbose=verbose, app=app, prelaunch_actions=prelaunch)
  File "/Users/peter.harrison/gdrive/academic/projects/harrison-peter/Dallinger/dallinger/command_line/docker.py", line 234, in deploy_heroku_docker
    launch_data = _handle_launch_data(launch_url, error=log)
  File "/Users/peter.harrison/gdrive/academic/projects/harrison-peter/Dallinger/dallinger/deployment.py", line 40, in _handle_launch_data
    launch_request = requests.post(url)
  File "/Users/peter.harrison/.virtualenvs/psynet/lib/python3.9/site-packages/requests/api.py", line 119, in post
    return request('post', url, data=data, json=json, **kwargs)
  File "/Users/peter.harrison/.virtualenvs/psynet/lib/python3.9/site-packages/requests/api.py", line 61, in request
    return session.request(method=method, url=url, **kwargs)
  File "/Users/peter.harrison/.virtualenvs/psynet/lib/python3.9/site-packages/requests/sessions.py", line 542, in request
    resp = self.send(prep, **send_kwargs)
  File "/Users/peter.harrison/.virtualenvs/psynet/lib/python3.9/site-packages/requests/sessions.py", line 655, in send
    r = adapter.send(request, **kwargs)
  File "/Users/peter.harrison/.virtualenvs/psynet/lib/python3.9/site-packages/requests/adapters.py", line 498, in send
    raise ConnectionError(err, request=request)
requests.exceptions.ConnectionError: ('Connection aborted.', RemoteDisconnected('Remote end closed connection without response'))
```

However when I called the launch route manually I could then access the app. Not sure if the manual launch route call was strictly necessary, sorry, I didn't test that.

- An important feature for reproducibility would be to export the Docker image as a permanent archive for the experiment. I know the images are stored locally and you can see them in `docker images`. It would be great though to think of a workflow for the experimenter that somehow keeps track of these, or gives some option for archiving in a remote repository, or similar. I'm not sure what the best solution is here and maybe it's better to experiment first before committing to anything. However I'd be interested in hearing any ideas you have.

I want to reiterate my excitement at the speed of deploying using Docker, once the image is cached. Now by far the bottleneck becomes waiting for Heroku postgres/redis/papertrail. I'm excited therefore about the plans to containerise these services too - this would massively speed up the process.

silviot commented:
@pmcharrison the issue you had should be solved now.
To get output during the push to heroku you need the verbose option (`dallinger docker sandbox --verbose`).

I added a configuration key (`image_base_name`) and two new commands `dallinger docker build` and `dallinger docker push`.

The idea is that a researcher can push the image to a docker registry and get an image name back, like

```
ghcr.io/dallinger/bartlett1932@sha256:165a645ab3385135c585902436819970577d91c1666ccab23fa56bce62f30f67
```
Next I'll work on #2807

silviot commented:
Closing in favour of #2863
